### PR TITLE
fix the cow_pow_ptr.h

### PR DESCRIPTION
**Pull request by lambda7xx**
fix the cow_pow_ptr.h  in order to build the utils

### CODE REVIEW

File: lib/utils/src/graph/serialparallel_internal.h
Comment (lockshaw) [inline lib/utils/src/graph/serialparallel_internal.h:1]: Remove from this PR

File: lib/utils/include/utils/graph/cow_ptr_t.h [hunk 1]
```diff
@@ -90,21 +95,33 @@ struct cow_ptr_t {
     swap(lhs.ptr, rhs.ptr);
   }
 private:
-  void set_shared(shared_t ptr) {
-    this->ptr = variant<shared_t>(std::move(ptr));
+  void set_shared(shared_t ptr) const {
```
Comment (lockshaw) [inline lib/utils/include/utils/graph/cow_ptr_t.h:98]: Why this change?

File: lib/utils/include/utils/graph/cow_ptr_t.h [hunk 2]
```diff
@@ -90,21 +95,33 @@ struct cow_ptr_t {
...
+  if (auto ptr = mpark::get_if<unique_t>(&this->ptr)) {
+    return std::move(*ptr);
   }
+  return nullptr;
+}
 
-  std::shared_ptr<T const> &get_shared() const {
-    return get<shared_t>(this->ptr);    
+ std::shared_ptr<T const> get_shared() const {
+  if (auto ptr = mpark::get_if<shared_t>(&this->ptr)) {
```
Comment (lockshaw) [inline lib/utils/include/utils/graph/cow_ptr_t.h:120]: Why change to returning `nullptr`? These are really only intended to be called when it has already been checked that they have the value (similar to `optional`)

File: lib/utils/include/utils/graph/cow_ptr_t.h (hunk 1 above)
Comment (lambda7xx) [reply lib/utils/include/utils/graph/cow_ptr_t.h:98]: because type of this->ptr is `mutable variant<std::unique_ptr<T>,std::shared_ptr<T const>> `

File: lib/utils/include/utils/graph/cow_ptr_t.h (hunk 2 above)
Comment (lambda7xx) [reply lib/utils/include/utils/graph/cow_ptr_t.h:120]: OK, I will fix it
